include/boost/geometry/algorithms/detail/partition.hpp

```diff
@@ -589,53 +653,57 @@ public :
 
     template
     <
-        typename InputCollection1,
-        typename InputCollection2,
+        typename ForwardRange1,
+        typename ForwardRange2,
         typename VisitPolicy
     >
-    static inline void apply(InputCollection1 const& collection1,
-                InputCollection2 const& collection2,
+    static inline void apply(ForwardRange1 const& forward_range1,
+                ForwardRange2 const& forward_range2,
                 VisitPolicy& visitor,
                 std::size_t min_elements = 16,
-                VisitBoxPolicy box_visitor = detail::partition::visit_no_policy()
+                VisitBoxPolicy box_visitor
+                    = detail::partition::visit_no_policy()
                 )
     {
-        if (std::size_t(boost::size(collection1)) > min_elements
-            && std::size_t(boost::size(collection2)) > min_elements)
+        typedef typename boost::range_iterator
+            <
+                ForwardRange1 const
+            >::type iterator_type1;
+
+        typedef typename boost::range_iterator
+            <
+                ForwardRange2 const
+            >::type iterator_type2;
+
+        if (std::size_t(boost::size(forward_range1)) > min_elements
+            && std::size_t(boost::size(forward_range2)) > min_elements)
         {
-            index_vector_type index_vector1, index_vector2;
+            std::vector<iterator_type1> iterator_vector1;
+            std::vector<iterator_type2> iterator_vector2;
```

Comment:
One more update on this: on my computer the size of `std::size_t` and the size of iterators of, for example, `bg::multi_point<>` and `bg::multi_linestring<>`, are both 8 bytes.
So my understanding is that for all existing usages, there is no overhead in using a vector of iterators instead of a vector of `std::size_t`'s.